## 회귀식 - 데이터가 적을 때 예측을 위하여 사용, 데이터가 적을 때 머신러닝을 돌리면    
## 머신러닝 - 데이터가 많을 때 예측을 위하여 사용     

#### Q 데이터가 많고 적음의 기준은 어느정도인가??     
5만개 정도로 생각하면 된다.     
아이리스 같이 데이터 정리가 잘 되어있는 데이터라면 100개라도 괜찮다.    
하지만 데이터가 분산되어 있는 데이터의 경우에 머신러닝을 돌리려면     
많은 데이터가 필요함.    
   
머신러닝 판단 - 아이리스 데이터의 경우 - 종 데이터가 몇퍼센트로 잘 분포하고 있나 잘 봐야한다.      

- 데이터를 어떻게 다룰줄을 모른다. 데이터를 알아야 분석을 할 수 있다.     


#### Q1 문제1
암환자 데이터 100개 중에 90개는 정상이고 10개는 암환자이다.   
라고 했을 때 어떻게 데이터를 나눠야 할까?   
현재 가지고 있는 데이터에서 현황을 분석하였을 때   
90퍼센트는 정상이고 , 10퍼센트가 암환자였다면    
머신러닝을 사용할 때도 데이터의 90퍼센트는 정상의 데이터   
10퍼센트는 암환자의 데이터를 가지고 머신러닝을     
적용시켜야 한다.   

#### Q2 문제2
헤드헌터가 취업자 수 데이터에서 100개중 80개는 미취업이고 20개는 취업일때   
어떻게 데이터를 나눠야 할까?    
위와같이 데이터의 비율을 나눠서 머신러닝 작업을 해야한다.   

분류 정제가 모두 끝났을 때 어떤흐름으로 작업을 시작해야 할까?    
데이터 다중공정성? 을 확인하여 데이터를 표준화 시킨다.    
데이터를 문제에 맞는 비율로 샘플링하고   
학습을 시키고     
예측 검증  
예측률 
전체 데이터로 학습  
결과를 적용시켜서 다시학습? (점점 데이터가 많아지도록?)      

In [2]:
data(iris)
summary(iris)






  Sepal.Length    Sepal.Width     Petal.Length    Petal.Width   
 Min.   :4.300   Min.   :2.000   Min.   :1.000   Min.   :0.100  
 1st Qu.:5.100   1st Qu.:2.800   1st Qu.:1.600   1st Qu.:0.300  
 Median :5.800   Median :3.000   Median :4.350   Median :1.300  
 Mean   :5.843   Mean   :3.057   Mean   :3.758   Mean   :1.199  
 3rd Qu.:6.400   3rd Qu.:3.300   3rd Qu.:5.100   3rd Qu.:1.800  
 Max.   :7.900   Max.   :4.400   Max.   :6.900   Max.   :2.500  
       Species  
 setosa    :50  
 versicolor:50  
 virginica :50  
                
                
                

### 표준화 작업 : 모든 입력변수의 평균값을 0으로 표준편차를 1로 정규화 한다
### 품종 컬럼을 제외하고 표준화 시키고 이를 품종 컬럼과 연결

In [38]:
# 표준화 작업 -- 다중공선성 때문에 표준화를 시킴
iris.scaled <- cbind(scale(iris[-5]),iris[5])
head(iris.scaled)
summary(iris.scaled)

# 표준편차
apply(iris.scaled[-5], 2, sd)









Sepal.Length Sepal.Width Petal.Length Petal.Width Species
1 -0.8976739    1.01560199 -1.335752    -1.311052   setosa 
2 -1.1392005   -0.13153881 -1.335752    -1.311052   setosa 
3 -1.3807271    0.32731751 -1.392399    -1.311052   setosa 
4 -1.5014904    0.09788935 -1.279104    -1.311052   setosa 
5 -1.0184372    1.24503015 -1.335752    -1.311052   setosa 
6 -0.5353840    1.93331463 -1.165809    -1.048667   setosa

  Sepal.Length       Sepal.Width       Petal.Length      Petal.Width     
 Min.   :-1.86378   Min.   :-2.4258   Min.   :-1.5623   Min.   :-1.4422  
 1st Qu.:-0.89767   1st Qu.:-0.5904   1st Qu.:-1.2225   1st Qu.:-1.1799  
 Median :-0.05233   Median :-0.1315   Median : 0.3354   Median : 0.1321  
 Mean   : 0.00000   Mean   : 0.0000   Mean   : 0.0000   Mean   : 0.0000  
 3rd Qu.: 0.67225   3rd Qu.: 0.5567   3rd Qu.: 0.7602   3rd Qu.: 0.7880  
 Max.   : 2.48370   Max.   : 3.0805   Max.   : 1.7799   Max.   : 1.7064  
       Species  
 setosa    :50  
 versicolor:50  
 virginica :50  
                
                
                

Sepal.Length  Sepal.Width Petal.Length  Petal.Width 
           1            1            1            1

In [39]:
# 학습용과 검증용 데이터셋을  7 : 3   /  데이터가 많을 땐 5 : 5 까지
#  150 : 105 (70%) + 45 (30%)
set.seed(1) # 샘플값 고정
samp <- c(sample(1:50, 35), sample(51:100, 35),sample(101:150, 35))

# 학습용 데이터
data.tr <- iris.scaled[samp,]
# 검증용 데이터
data.test <- iris.scaled[-samp,]

summary(data.tr)
summary(data.test)









  Sepal.Length       Sepal.Width         Petal.Length       Petal.Width       
 Min.   :-1.86378   Min.   :-2.425820   Min.   :-1.56234   Min.   :-1.442245  
 1st Qu.:-1.01844   1st Qu.:-0.590395   1st Qu.:-1.27910   1st Qu.:-1.179860  
 Median :-0.17309   Median :-0.131539   Median : 0.30703   Median : 0.132067  
 Mean   :-0.08568   Mean   :-0.000437   Mean   :-0.02368   Mean   :-0.004123  
 3rd Qu.: 0.67225   3rd Qu.: 0.556746   3rd Qu.: 0.76021   3rd Qu.: 0.788031  
 Max.   : 2.48370   Max.   : 2.621599   Max.   : 1.66657   Max.   : 1.706379  
       Species  
 setosa    :35  
 versicolor:35  
 virginica :35  
                
                
                

  Sepal.Length       Sepal.Width        Petal.Length       Petal.Width       
 Min.   :-1.13920   Min.   :-1.96696   Min.   :-1.44905   Min.   :-1.311052  
 1st Qu.:-0.53538   1st Qu.:-0.59040   1st Qu.:-1.16581   1st Qu.:-1.048667  
 Median : 0.06843   Median :-0.13154   Median : 0.36368   Median : 0.132067  
 Mean   : 0.19993   Mean   : 0.00102   Mean   : 0.05526   Mean   : 0.009621  
 3rd Qu.: 0.79301   3rd Qu.: 0.55675   3rd Qu.: 0.87351   3rd Qu.: 0.788031  
 Max.   : 2.24217   Max.   : 3.08046   Max.   : 1.77987   Max.   : 1.706379  
       Species  
 setosa    :15  
 versicolor:15  
 virginica :15  
                
                
                

In [40]:
library(nnet) # 인공신경망 라이브러리





In [41]:
#  nnet (종속변수 ~ 독립변수), size : 은닉층갯수
model.nnet <- nnet(Species ~ . , data = data.tr, size = 3)
model.nnet





# weights:  27
initial  value 123.783256 
iter  10 value 9.852981
iter  20 value 2.073465
iter  30 value 0.003033
final  value 0.000039 
converged


a 4-3-3 network with 27 weights
inputs: Sepal.Length Sepal.Width Petal.Length Petal.Width 
output(s): Species 
options were - softmax modelling 

initial  value 123.783256     
iter  10 value 9.852981   
iter  20 value 2.073465    
iter  30 value 0.003033   
final  value 0.000039     -->  최종학습 결과 / 최소한 20회 이상 학습한것?  


In [42]:
 # 신경망 모델의 검증
# predict(), class
predicted <- predict(model.nnet, data.test, type = "class")
predicted
# 이 값을 수치화 시켜야 함
# Classfication 분류의 약자





[1] "setosa"     "setosa"     "setosa"     "setosa"     "setosa"    
 [6] "setosa"     "setosa"     "setosa"     "setosa"     "setosa"    
[11] "setosa"     "setosa"     "setosa"     "setosa"     "setosa"    
[16] "versicolor" "versicolor" "versicolor" "versicolor" "versicolor"
[21] "versicolor" "versicolor" "versicolor" "versicolor" "versicolor"
[26] "versicolor" "versicolor" "versicolor" "versicolor" "versicolor"
[31] "virginica"  "virginica"  "virginica"  "virginica"  "virginica" 
[36] "virginica"  "virginica"  "virginica"  "virginica"  "virginica" 
[41] "virginica"  "virginica"  "versicolor" "virginica"  "virginica"

In [43]:
# 검증용 데이터셋의 예측결과를 벡터의 형태로 변형
actual <- data.test$Species # 정답 변수
result <- table(actual,predicted)
result

# 각 항목의 맞춘 갯수가 나온다





            predicted
actual       setosa versicolor virginica
  setosa         15          0         0
  versicolor      0         15         0
  virginica       0          1        14

In [44]:
# 모델의 정확도
paste("예측확률",round(mean(predicted == actual),2) * 100 ,"%")





[1] "예측확률 98 %"

# 아이리스 데이터전체로 

리스트로 값을 줘야 학습시킬 수 있음

In [116]:
# 표준화 작업 없애기
iris.scaled <- iris


# 표준편차
apply(iris.scaled[-5], 2, sd)


library(nnet) # 인공신경망 라이브러리

#  nnet (종속변수 ~ 독립변수), size : 은닉층갯수
model.nnet <- nnet(Species ~ . , data = iris.scaled, size = 3)
model.nnet

 # 신경망 모델의 검증
# predict(), class
zz <- list(Sepal.Length=-0.5514857, Sepal.Width=-0.1315388, Petal.Length=1.1567451, Petal.Width=1.4439941)
predicted <- predict(model.nnet, zz, type = "class")
predicted
# 이 값을 수치화 시켜야 함

# 검증용 데이터셋의 예측결과를 벡터의 형태로 변형
actual <- data.test$Species # 정답 변수


# 각 항목의 맞춘 갯수가 나온다

# 모델의 정확도
paste("예측확률",round(mean(predicted == actual),2) * 100 ,"%")













Sepal.Length  Sepal.Width Petal.Length  Petal.Width 
   0.8280661    0.4358663    1.7652982    0.7622377

# weights:  27
initial  value 176.425005 
iter  10 value 21.672904
iter  20 value 7.091795
iter  30 value 5.745370
iter  40 value 4.946995
iter  50 value 4.922932
iter  60 value 4.921967
final  value 4.921966 
converged


a 4-3-3 network with 27 weights
inputs: Sepal.Length Sepal.Width Petal.Length Petal.Width 
output(s): Species 
options were - softmax modelling 

[1] "versicolor"

[1] "예측확률 33 %"

# 데이터를 5대 5비율로

# 한셀로 과정을 정리하고 이해하기

In [119]:
print("================================================= 표준화")
# 표준화 작업 - 다중 공선성을 지키려 표준화를 시킨다
iris.scaled <- cbind(scale(iris[-5]),iris[5])   # 표준화 시킨 데이터를 다시 데이터 프레임 만들기
head(iris.scaled)
summary(iris.scaled)

# 표준편차
apply(iris.scaled[-5], 2, sd)

print("================================================= 데이터셋")
# 학습용과 검증용 데이터셋 나누기
#  150 : 105 (70%) + 45 (30%)
set.seed(1) # 샘플값 고정
samp <- c(sample(1:50, 35), sample(51:100, 35),sample(101:150, 35))

# 학습용 데이터
data.tr <- iris.scaled[samp,]
# 검증용 데이터
data.test <- iris.scaled[-samp,]

summary(data.tr)
summary(data.test)

print("================================================= nnet")
#  nnet (종속변수 ~ 독립변수), size : 은닉층갯수
model.nnet <- nnet(Species ~ . , data = data.tr, size = 3)
model.nnet

print("================================================= predict")
 # 신경망 모델의 검증 predict(), class
predicted <- predict(model.nnet, data.test, type = "class")
predicted


print("================================================= 예측결과")
# 검증용 데이터셋의 예측결과를 벡터의 형태로 변형
actual <- data.test$Species # 정답 변수
result <- table(actual,predicted)
result


print("================================================= 정확도")
# 모델의 정확도
paste("예측확률",round(mean(predicted == actual),2) * 100 ,"%")

























[1] "================================================= 표준화"


Sepal.Length Sepal.Width Petal.Length Petal.Width Species
1 -0.8976739    1.01560199 -1.335752    -1.311052   setosa 
2 -1.1392005   -0.13153881 -1.335752    -1.311052   setosa 
3 -1.3807271    0.32731751 -1.392399    -1.311052   setosa 
4 -1.5014904    0.09788935 -1.279104    -1.311052   setosa 
5 -1.0184372    1.24503015 -1.335752    -1.311052   setosa 
6 -0.5353840    1.93331463 -1.165809    -1.048667   setosa

  Sepal.Length       Sepal.Width       Petal.Length      Petal.Width     
 Min.   :-1.86378   Min.   :-2.4258   Min.   :-1.5623   Min.   :-1.4422  
 1st Qu.:-0.89767   1st Qu.:-0.5904   1st Qu.:-1.2225   1st Qu.:-1.1799  
 Median :-0.05233   Median :-0.1315   Median : 0.3354   Median : 0.1321  
 Mean   : 0.00000   Mean   : 0.0000   Mean   : 0.0000   Mean   : 0.0000  
 3rd Qu.: 0.67225   3rd Qu.: 0.5567   3rd Qu.: 0.7602   3rd Qu.: 0.7880  
 Max.   : 2.48370   Max.   : 3.0805   Max.   : 1.7799   Max.   : 1.7064  
       Species  
 setosa    :50  
 versicolor:50  
 virginica :50  
                
                
                

Sepal.Length  Sepal.Width Petal.Length  Petal.Width 
           1            1            1            1

[1] "================================================= 데이터셋"


  Sepal.Length       Sepal.Width         Petal.Length       Petal.Width       
 Min.   :-1.86378   Min.   :-2.425820   Min.   :-1.56234   Min.   :-1.442245  
 1st Qu.:-1.01844   1st Qu.:-0.590395   1st Qu.:-1.27910   1st Qu.:-1.179860  
 Median :-0.17309   Median :-0.131539   Median : 0.30703   Median : 0.132067  
 Mean   :-0.08568   Mean   :-0.000437   Mean   :-0.02368   Mean   :-0.004123  
 3rd Qu.: 0.67225   3rd Qu.: 0.556746   3rd Qu.: 0.76021   3rd Qu.: 0.788031  
 Max.   : 2.48370   Max.   : 2.621599   Max.   : 1.66657   Max.   : 1.706379  
       Species  
 setosa    :35  
 versicolor:35  
 virginica :35  
                
                
                

  Sepal.Length       Sepal.Width        Petal.Length       Petal.Width       
 Min.   :-1.13920   Min.   :-1.96696   Min.   :-1.44905   Min.   :-1.311052  
 1st Qu.:-0.53538   1st Qu.:-0.59040   1st Qu.:-1.16581   1st Qu.:-1.048667  
 Median : 0.06843   Median :-0.13154   Median : 0.36368   Median : 0.132067  
 Mean   : 0.19993   Mean   : 0.00102   Mean   : 0.05526   Mean   : 0.009621  
 3rd Qu.: 0.79301   3rd Qu.: 0.55675   3rd Qu.: 0.87351   3rd Qu.: 0.788031  
 Max.   : 2.24217   Max.   : 3.08046   Max.   : 1.77987   Max.   : 1.706379  
       Species  
 setosa    :15  
 versicolor:15  
 virginica :15  
                
                
                

[1] "================================================= nnet"
# weights:  27
initial  value 123.783256 
iter  10 value 9.852981
iter  20 value 2.073465
iter  30 value 0.003033
final  value 0.000039 
converged


a 4-3-3 network with 27 weights
inputs: Sepal.Length Sepal.Width Petal.Length Petal.Width 
output(s): Species 
options were - softmax modelling 

[1] "================================================= predict"


[1] "setosa"     "setosa"     "setosa"     "setosa"     "setosa"    
 [6] "setosa"     "setosa"     "setosa"     "setosa"     "setosa"    
[11] "setosa"     "setosa"     "setosa"     "setosa"     "setosa"    
[16] "versicolor" "versicolor" "versicolor" "versicolor" "versicolor"
[21] "versicolor" "versicolor" "versicolor" "versicolor" "versicolor"
[26] "versicolor" "versicolor" "versicolor" "versicolor" "versicolor"
[31] "virginica"  "virginica"  "virginica"  "virginica"  "virginica" 
[36] "virginica"  "virginica"  "virginica"  "virginica"  "virginica" 
[41] "virginica"  "virginica"  "versicolor" "virginica"  "virginica"

[1] "================================================= 예측결과"


            predicted
actual       setosa versicolor virginica
  setosa         15          0         0
  versicolor      0         15         0
  virginica       0          1        14

[1] "================================================= 정확도"


[1] "예측확률 98 %"

--------------------
# 궁금점 해소하기!

In [188]:
# 난수고정은 왜 하는 것일까?
set.seed(1) 
samp <- c(sample(1:50, 35), sample(51:100, 35),sample(101:150, 35))
samp





[1]  14  19  28  43  10  41  42  29  27   3   9   7  44  15  48  18  25  33
 [19]  13  34  47  39  49   4  30  46   1  40  20   8  31  12  23  38  50  84
 [37]  89  56  85  69  87  79 100  74  72  82  51  96  78  75  67  80  65  58
 [55]  53  81  60  83  68  61  73  90  76  92  64  98  94  71  52  95 117 142
 [73] 150 116 122 141 139 148 133 140 118 128 147 113 144 108 125 105 135 149
 [91] 132 102 145 124 121 120 111 110 136 137 114 107 131 143 127

In [128]:
cor(iris[-5]) # 표준화 시켜야겠다





Sepal.Length Sepal.Width Petal.Length Petal.Width
Sepal.Length  1.0000000   -0.1175698   0.8717538    0.8179411 
Sepal.Width  -0.1175698    1.0000000  -0.4284401   -0.3661259 
Petal.Length  0.8717538   -0.4284401   1.0000000    0.9628654 
Petal.Width   0.8179411   -0.3661259   0.9628654    1.0000000

In [130]:
# 표준화 시키기
iris.scaled <- cbind(scale(iris[-5]),iris[5])





In [ ]:
# 데이터 분류 시키기
# 학습용 데이터
data.tr <- iris.scaled[samp,]
# 검증용 데이터
data.test <- iris.scaled[-samp,]





In [150]:
# 머신러닝은 실핼할 때 마다 데이터 변경! 
library(nnet)
model.nnet <- nnet(Species ~ . , data = data.tr, size = 6)
model.nnet





# weights:  51
initial  value 123.823332 
iter  10 value 13.630250
iter  20 value 2.703851
iter  30 value 0.125119
iter  40 value 0.000568
final  value 0.000083 
converged


a 4-6-3 network with 51 weights
inputs: Sepal.Length Sepal.Width Petal.Length Petal.Width 
output(s): Species 
options were - softmax modelling 

In [154]:
 # 신경망 모델의 검증 predict(), class
predicted <- predict(model.nnet, data.test, type = "class")

# 검증용 데이터셋의 예측결과를 벡터의 형태로 변형
actual <- data.test$Species # 정답 변수
result <- table(actual,predicted)
table(actual,predicted)
mean(actual == predicted) * 100





            predicted
actual       setosa versicolor virginica
  setosa         15          0         0
  versicolor      0         15         0
  virginica       0          2        13

[1] 95.55556

In [169]:
# 전체 데이터 학습
model2.nnet <- nnet(Species ~ . , data = iris.scaled, size = 10)





# weights:  83
initial  value 177.742311 
iter  10 value 12.059665
iter  20 value 5.510348
iter  30 value 3.500997
iter  40 value 2.197450
iter  50 value 1.903399
iter  60 value 1.670974
iter  70 value 0.016748
iter  80 value 0.000277
final  value 0.000060 
converged


### 도데체 set.seed는 왜 한걸까?
set.seed는 난수를 뽑는 로직이 있는데 그것 조차 고정시켜 준다.
만약 1:100 까지 setosa 데이터가 1:1000 까지로 변화 되었을 때 다시 난수를 뽑는다 해도 그 로직이 고정되어 ㅣ
데이터가 새로 추가될 때 rbind를 하고 csv로 들어갔다 오면 인덱스가 변화되어 있다.
이렇게 되면 샘플링 번호는 
그래서 데이터가 섞이